libs/data-access/lessons/src/lib/saga.ts

```diff
@@ -0,0 +1,163 @@
+import { call, put, takeEvery } from 'redux-saga/effects';
+import { ActionTypes } from './constants';
+import {
+  getAllSuccess,
+  getAllFailed,
+  createSuccess,
+  createFailed,
+  updateSuccess,
+  removeSuccess,
+  removeFailed,
+  getByIdSuccess,
+  getByIdFailed,
+} from './actions';
+import { post, postFormData, get } from '@ppm/data-access/http-requests';
+import { PrivateRoutesPath } from '@ppm/common/main';
+
+export function* createLesson(actions) {
+  const data = actions.payload;
+  try {
+    if (data.lessonImage.length) {
+      const file = data.lessonImage[0];
+      const formData = new FormData();
+      formData.append('file', file);
+      const path = `/api/${PrivateRoutesPath.IMAGES}`;
+      const imageResult = yield call(postFormData, path, formData);
+
+      if (imageResult) {
+        const path = `/api/${PrivateRoutesPath.LESSONS}`;
+        const data = actions.payload;
+        yield call(post, path, data);
+        yield put(
+          createSuccess({
+            loading: false,
+          })
+        );
+      }
+    } else {
+      const path = `/api/${PrivateRoutesPath.LESSONS}`;
+      const data = actions.payload;
+      yield call(post, path, data);
+      yield put(
+        createSuccess({
+          loading: false,
+        })
+      );
+    }
+  } catch (error) {
+    yield put(createFailed());
+  }
+}
+
+export function* updateLesson(actions) {
+  const data = actions.payload;
+  try {
+    if (data.lessonImage.length) {
+      const file = data.lessonImage[0];
+      const formData = new FormData();
+      formData.append('file', file);
+      const path = `/api/${PrivateRoutesPath.IMAGES}`;
+      const imageResult = yield call(postFormData, path, formData);
+
+      if (imageResult) {
+        const path = `/api/${PrivateRoutesPath.LESSONS}/update/${data.id}`;
+        yield call(post, path, { ...data, imageUrl: imageResult.data });
+        yield put(
+          updateSuccess({
+            loading: false,
+          })
+        );
+      }
+    } else {
+      const path = `/api/${PrivateRoutesPath.LESSONS}/update/${data.id}`;
+      yield call(post, path, data);
+      yield put(
+        updateSuccess({
+          loading: false,
+        })
+      );
+    }
+  } catch (error) {}
+}
+
+export function* removeLesson(actions) {
+  try {
+    const id = actions.payload;
+    const path = `/api/${PrivateRoutesPath.LESSONS}/delete/${id}`;
+    const result = yield call(post, path);
+    if (result) {
+      yield put(removeFailed());
+    }
+    yield put(removeSuccess());
+  } catch (error) {
+    yield put(removeFailed());
+  }
+}
+
+export function* getLessonById(actions) {
+  try {
+    const id = actions.payload;
+    const path = `/api/${PrivateRoutesPath.LESSONS}/${PrivateRoutesPath.USER}/${id}`;
+    const result = yield call(get, path);
+
+    if (!!result && !Array.isArray(result.data)) {
```

Comment:
this no need double !! this operator required only in component html part, same below